include/mbgl/map/map_observer.hpp

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+namespace mbgl {
+
+class MapObserver {
+public:
+    static MapObserver& nullObserver() {
+        static MapObserver mapObserver;
+        return mapObserver;
+    }
+
+    virtual void onRegionWillChange() {}
+    virtual void onRegionWillChangeAnimated() {}
+    virtual void onRegionIsChanging() {}
+    virtual void onRegionDidChange() {}
+    virtual void onRegionDidChangeAnimated() {}
+    virtual void onWillStartLoadingMap() {}
+    virtual void onDidFinishLoadingMap() {}
+    virtual void onDidFailLoadingMap() {}
+    virtual void onWillStartRenderingFrame() {}
+    virtual void onDidFinishRenderingFrame() {}
+    virtual void onDidFinishRenderingFrameFullyRendered() {}
+    virtual void onWillStartRenderingMap() {}
```

Comment:
In terms of usefulness, we’ve found the “rendering frame” notifications to be useful in situations where the developer wants to synchronize a view atop the map or override some annotation view behavior that we apply on every frame. But the “rendering map” notifications remain more generally useful for knowing when it’s safe to snapshot a map view or have the UI change once the map is quiescent.